securedrop_client/gui/widgets.py

```diff
@@ -849,20 +849,37 @@ def update(self, sources: List[Source]):
         current_source_id = current_source and current_source.id
 
         self.clear()
-
-        for source in sources:
-            new_source = SourceWidget(source)
-            new_source.setup(self.controller)
-            self.source_widgets[source.uuid] = new_source
-
-            list_item = QListWidgetItem(self)
-            list_item.setSizeHint(new_source.sizeHint())
-
-            self.addItem(list_item)
-            self.setItemWidget(list_item, new_source)
-
-            if source.id == current_source_id and has_source_selected:
-                self.setCurrentItem(list_item)
+        sources.reverse()
+        self.add_source(sources, current_source_id, has_source_selected)
+
+    def add_source(self, sources, current_source_id, has_source_selected, slice_size=1):
+        """
+        Add a slice of sources, ensure the currently selected source is
+        retained and, if necessary, reschedule the addition of more sources.
+        """
+        def schedule_source_creation(slice_size=slice_size):
+            if sources:
+                sources_slice = sources[:slice_size]  # arbitrary
+                for source in sources_slice:
+                    new_source = SourceWidget(source)
+                    new_source.setup(self.controller)
+                    self.source_widgets[source.uuid] = new_source
```

Comment:
Yeah... I noticed this too, but wanted this to be a nicely constrained PR. If we have a huge number of sources, this will not look good unless we do something like in the `ConversationView` (update existing widgets, rather than recreate the universe on each update from the API -- potentially every 15 seconds).

Reply:
totally, now being worked on in #893